File: WEEK6/python/bj8_1256.py
import sys

input = sys.stdin.readline

n, m, k = map(int, input().split())
dp = [[0] * 201 for _ in range(201)]
limit = 1000000000

for i in range(201):
    dp[i][0] = 1
    for j in range(1, i + 1):
        dp[i][j] = dp[i - 1][j - 1] + dp[i - 1][j]
        if dp[i][j] > limit:
            dp[i][j] = limit + 1

if dp[n + m][m] < k:
    print("-1")
else:
    result = []
    while n > 0 or m > 0:
        a_count = dp[n + m - 1][m] if n > 0 else 0
        if n > 0 and k <= a_count:
            result.append("a")
            n -= 1
        else:
            result.append("z")
            k -= a_count
            m -= 1
    print("".join(result))

File: WEEK6/Js/bj8_1256.js
const fs = require('fs');
const [N, M, K_Input] = fs.readFileSync('/dev/stdin').toString().trim().split(' ').map(Number);

let n = N;
let m = M;
let k = BigInt(K_Input);

const dp = Array.from({ length: 201 }, () => new BigInt64Array(201));
const MAX = 1000000000n;

for (let i = 0; i <= 200; i++) {
    dp[i][0] = 1n;
    for (let j = 1; j <= i; j++) {
        dp[i][j] = dp[i - 1][j - 1] + dp[i - 1][j];
        if (dp[i][j] > MAX) dp[i][j] = MAX + 1n;
    }
}

if (dp[n + m][m] < k) {
    console.log("-1");
    process.exit(0);
}

let result = "";
while (n > 0 || m > 0) {
    let aCount = (n > 0) ? dp[n + m - 1][m] : 0n;

    if (n > 0 && k <= aCount) {
        result += "a";
        n--;
    } else {
        result += "z";
        k -= aCount;
        m--;
    }
}

console.log(result);